tests/linearizability/model.go

```diff
@@ -167,33 +240,41 @@ func applyRequest(states PossibleStates, request EtcdRequest, response EtcdRespo
 
 // applyRequestToSingleState handles a successful request, returning updated state and response it would generate.
 func applyRequestToSingleState(s EtcdState, request EtcdRequest) (EtcdState, EtcdResponse) {
+	success := true
+	for _, cond := range request.Conds {
+		if val := s.KeyValues[cond.Key]; val != cond.ExpectedValue {
+			success = false
+		}
+	}
+	if !success {
+		return s, EtcdResponse{Revision: s.Revision, TxnFailure: true}
+	}
 	newKVs := map[string]string{}
 	for k, v := range s.KeyValues {
 		newKVs[k] = v
 	}
 	s.KeyValues = newKVs
-	resp := EtcdResponse{}
-	switch request.Op {
-	case Get:
-		resp.GetData = s.KeyValues[request.Key]
-	case Put:
-		s.KeyValues[request.Key] = request.PutData
-		s.Revision += 1
-	case Delete:
-		if _, ok := s.KeyValues[request.Key]; ok {
-			delete(s.KeyValues, request.Key)
-			s.Revision += 1
-			resp.Deleted = 1
-		}
-	case Txn:
-		if val := s.KeyValues[request.Key]; val == request.TxnExpectData {
-			s.KeyValues[request.Key] = request.TxnNewData
-			s.Revision += 1
-			resp.TxnSucceeded = true
+	opResp := make([]EtcdOperationResult, len(request.Ops))
+	increaseRevision := false
+	for i, op := range request.Ops {
+		switch op.Type {
+		case Get:
+			opResp[i].Value = s.KeyValues[op.Key]
+		case Put:
+			s.KeyValues[op.Key] = op.Value
+			increaseRevision = true
```

Comment:
The slice `opResp` isn't updated at all in this case (`Put`), it might not be correct. Let's work with an example, assuming there are three operations in one TXN,
- read
- put
- delete

The length of `opResp` will be 3, but only the first and probably third elements are populated, the second element is still `nil`. Accordingly it will not match the actual response, see https://github.com/etcd-io/etcd/blob/3b623afddd27c419f00fb312128af466031b5549/tests/linearizability/model.go#L234

The reason why the workflow is green is there is only one operation in each request for now. When you start to support multiple operations, it might be failed.

Reply:
It should be correct as Txn results are positional, elements are left empty intentionally.